File: 698.partition-to-k-equal-sum-subsets.py
#
# @lc app=leetcode id=698 lang=python3
#
# [698] Partition to K Equal Sum Subsets
#

# @lc code=start
from typing import *


class Solution:
    def canPartitionKSubsets(self, nums: List[int], k: int) -> bool:
        sum_val = sum(nums)
        if sum_val % k:
            return False
        self.length = len(nums)
        self.k = k
        self.required_num = sum_val//k
        nums.sort(reverse=True)
        subset = [0]*k
        return self.backtrack(subset, nums, 0)

    def backtrack(self, subset: List[int], nums: List[int], pos: int) -> bool:
        if pos == self.length:
            return True
        for i in range(self.k):
            seen = False
            for j in range(i):
                if subset[i] == subset[j]:
                    seen = True
                    break
            if seen or subset[i] + nums[pos] > self.required_num:
                continue
            subset[i] += nums[pos]
            if self.backtrack(subset, nums, pos + 1):
                return True
            subset[i] -= nums[pos]
        return False
# @lc code=end

File: 698.partition-to-k-equal-sum-subsets.cpp
/*
 * @lc app=leetcode id=698 lang=cpp
 *
 * [698] Partition to K Equal Sum Subsets
 */
#include <vector>
#include <numeric>
#include <algorithm>
using namespace std;
// @lc code=start
class Solution
{
    int length;

public:
    bool canPartitionKSubsets(vector<int> &nums, int k)
    {
        int sum_val = accumulate(nums.begin(), nums.end(), 0);
        if (sum_val % k)
        {
            return false;
        }
        length = nums.size();
        int required_num = sum_val / k;
        sort(nums.begin(), nums.end(), greater<int>());
        vector<int> subset(k);
        return backtrack(subset, k, required_num, nums, 0);
    }
    bool backtrack(vector<int> &subset, const int k, const int required_num, vector<int> &nums, int pos)
    {
        if (pos == length)
        {
            return true;
        }
        for (int i = 0; i != k; ++i)
        {
            bool seen = false;
            for (int j = 0; j != i; ++j)
            {
                if (subset[i] == subset[j])
                {
                    seen = true;
                    break;
                }
            }
            if (seen || subset[i] + nums[pos] > required_num)
            {
                continue;
            }
            subset[i] += nums[pos];
            if (backtrack(subset, k, required_num, nums, pos + 1))
            {
                return true;
            }
            subset[i] -= nums[pos];
        }
        return false;
    }
};
// @lc code=end
